### PR TITLE
Semantic syntax highlighting

### PR DESCRIPTION
**Pull request by kirawi**

### CODE REVIEW

File: helix-lsp/src/client.rs [hunk 1]
```diff
@@ -800,4 +846,18 @@ impl Client {
...
+        text_document: lsp::TextDocumentIdentifier,
+        work_done_token: Option<lsp::ProgressToken>,
+    ) -> Result<lsp::SemanticTokensResult> {
+        let params = lsp::SemanticTokensParams {
+            work_done_progress_params: lsp::WorkDoneProgressParams { work_done_token },
+            partial_result_params: lsp::PartialResultParams::default(),
+            text_document,
+        };
+        // Typo from lsp_types
+        self.request::<lsp::request::SemanticTokensRefesh>(params).await
```
Comment (kirawi) [inline helix-lsp/src/client.rs:861]: /cc @archseer How do I resolve the error here?
```
mismatched types
expected enum `std::result::Result<SemanticTokensResult, _>`
   found enum `std::result::Result<(), _>`
```
```
mismatched types
expected `()`, found struct `SemanticTokensParams`
```

**Comment by archseer** [comment]
I'm still not sure if semantic highlighting is something we want to support, it's a terrible spec.

> “But what about LSP?” I can hear some of you say. The reason (most) LSP servers don’t offer syntax highlighting is because of the drag on performance. Every keystroke you type must be sent to the server, processed, a partial tree returned, and your syntax highlighting updated. Repeat that up to 100 words per minute (or whatever your typing speed is) and you’re looking at a lot of cross-chatter that is just better suited for in-process communication. But of course that doesn’t mean it can’t replace the language parsing used for other features in LSP!

https://www.masteringemacs.org/article/tree-sitter-complications-of-parsing-languages

**Comment by kirawi** [comment]
I'll close this then and implement it on my branch until it can be made a plugin.

**Comment by kirawi** [comment]
It would be nice if we could replicate the benefits of semantic syntax higlighting through tree-sitter, but that sounds pretty complicated.

**Comment by archseer** [comment]
What are some of the benefits? Highlighting mutable variables can be done through `locals.scm`

**Comment by archseer** [comment]
Doc comments can be highlighted differently once https://github.com/tree-sitter/tree-sitter-rust/pull/99 is merged

**Comment by kirawi** [comment]
It doesn't look like it can omit the leading characters `///` which complicates it. Impossible was the wrong word to use, I meant that it is difficult. From https://github.com/tree-sitter/tree-sitter-c-sharp/pull/208 it looks that while it's possible to do doc comments, it's not as easy as injections.

**Pull request closed without merging**